File: CLDemo/Class/My/View/CLPasswordInputView/CLPasswordInputView.m
//
//  CLPasswordInputView.m
//  CLDemo
//
//  Created by AUG on 2019/1/15.
//  Copyright © 2019年 JmoVxia. All rights reserved.
//

#import "CLPasswordInputView.h"
#import "UIColor+CLHex.h"

@implementation CLPasswordInputViewConfigure

+ (instancetype)defaultConfig {
    CLPasswordInputViewConfigure *configure = [[CLPasswordInputViewConfigure alloc] init];
    configure.squareWidth = 50;
    configure.passwordNum = 6;
    configure.pointRadius = 18 * 0.5;
    configure.spaceMultiple = 5;
    configure.rectColor = [UIColor colorWithRGBHex:0xb2b2b2];
    configure.pointColor = [UIColor blackColor];
    configure.rectBackgroundColor = [UIColor whiteColor];
    configure.backgroundColor = [UIColor whiteColor];
    configure.threePartyKeyboard = NO;
    return configure;
}

@end

@interface CLPasswordInputView ()

@property (nonatomic, strong) CLPasswordInputViewConfigure *configure;

@property (nonatomic, strong) NSMutableString *text;

@property (nonatomic, assign) BOOL isShow;

@end


@implementation CLPasswordInputView

- (CLPasswordInputViewConfigure *) configure{
    if (_configure == nil){
        _configure = [CLPasswordInputViewConfigure defaultConfig];
    }
    return _configure;
}
- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.text = [NSMutableString string];
        self.backgroundColor = self.configure.backgroundColor;
    }
    return self;
}
- (void)clearText {
    self.text = [NSMutableString string];
    [self setNeedsLayout];
}
- (UIKeyboardType)keyboardType {
    return UIKeyboardTypeNumberPad;
}
- (BOOL)becomeFirstResponder {
    if (!self.isShow) {
        if ([self.delegate respondsToSelector:@selector(passwordInputViewBeginInput:)]) {
            [self.delegate passwordInputViewBeginInput:self];
        }
    }
    self.isShow = YES;
    return [super becomeFirstResponder];
}
- (BOOL)resignFirstResponder {
    if (self.isShow) {
        if ([self.delegate respondsToSelector:@selector(passwordInputViewEndInput:)]) {
            [self.delegate passwordInputViewEndInput:self];
        }
    }
    self.isShow = NO;
    return [super resignFirstResponder];
}
- (BOOL)canBecomeFirstResponder {
    return YES;
}
- (BOOL)canResignFirstResponder {
    return YES;
}
- (BOOL)isSecureTextEntry {
    return !self.configure.threePartyKeyboard;
}
- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if (![self isFirstResponder]) {
        [self becomeFirstResponder];
    }
}
- (void)updateWithConfigure:(void(^)(CLPasswordInputViewConfigure *configure))configBlock {
    if (configBlock) {
        configBlock(self.configure);
    }
    self.backgroundColor = self.configure.backgroundColor;
    [self setNeedsDisplay];
}
//MARK:JmoVxia---UIKeyInput
- (BOOL)hasText {
    return self.text.length > 0;
}
- (void)insertText:(NSString *)text {
    if (self.text.length < self.configure.passwordNum) {
        //判断是否是数字
        NSCharacterSet *cs = [[NSCharacterSet characterSetWithCharactersInString:@"0123456789"] invertedSet];
        NSString*filtered = [[text componentsSeparatedByCharactersInSet:cs] componentsJoinedByString:@""];
        BOOL basicTest = [text isEqualToString:filtered];
        if(basicTest) {
            [self.text appendString:text];
            if ([self.delegate respondsToSelector:@selector(passwordInputViewDidChange:)]) {
                [self.delegate passwordInputViewDidChange:self];
            }
            if (self.text.length == self.configure.passwordNum) {
                if ([self.delegate respondsToSelector:@selector(passwordInputViewCompleteInput:)]) {
                    [self.delegate passwordInputViewCompleteInput:self];
                }
            }
            [self setNeedsDisplay];
        }
    }
}
- (void)deleteBackward {
    if (self.text.length > 0) {
        [self.text deleteCharactersInRange:NSMakeRange(self.text.length - 1, 1)];
        if ([self.delegate respondsToSelector:@selector(passwordInputViewDidChange:)]) {
            [self.delegate passwordInputViewDidChange:self];
        }
    }
    if ([self.delegate respondsToSelector:@selector(passwordInputViewDidDeleteBackward:)]) {
        [self.delegate passwordInputViewDidDeleteBackward:self];
    }
    [self setNeedsDisplay];
}

-(void)layoutSubviews {
    [super layoutSubviews];
    [self setNeedsDisplay];
}

- (void)drawRect:(CGRect)rect {
    CGFloat height = rect.size.height;
    CGFloat width = rect.size.width;
    CGFloat squareWidth = MIN(MAX(MIN(height, self.configure.squareWidth), (self.configure.pointRadius * 4)), height);
    CGFloat pointRadius = MIN(self.configure.pointRadius, squareWidth * 0.5) * 0.8;
    CGFloat middleSpace = (width - self.configure.passwordNum * squareWidth) / (self.configure.passwordNum - 1 + self.configure.spaceMultiple * 2);
    CGFloat leftSpace = middleSpace * self.configure.spaceMultiple;
    CGFloat y = (height - squareWidth) * 0.5;
    CGContextRef context = UIGraphicsGetCurrentContext();
    //画外框
    for (NSUInteger i = 0; i < self.configure.passwordNum; i++) {
        CGContextAddRect(context, CGRectMake(leftSpace + i * squareWidth + i * middleSpace, y, squareWidth, squareWidth));
        CGContextSetLineWidth(context, 1);
        CGContextSetStrokeColorWithColor(context, self.configure.rectColor.CGColor);
        CGContextSetFillColorWithColor(context, self.configure.rectBackgroundColor.CGColor);
    }
    CGContextDrawPath(context, kCGPathFillStroke);
    CGContextSetFillColorWithColor(context, self.configure.pointColor.CGColor);
    //画黑点
    for (NSUInteger i = 1; i <= self.text.length; i++) {
        CGContextAddArc(context,  leftSpace + i * squareWidth + (i - 1) * middleSpace - squareWidth * 0.5, y + squareWidth * 0.5, pointRadius, 0, M_PI * 2, YES);
        CGContextDrawPath(context, kCGPathFill);
    }
}

@end

File: CLDemo-Swift/CLDemo-Swift/Class/CLPasswordController/CLPasswordInputView.swift
//
//  CLPasswordInputView.swift
//  CLDemo
//
//  Created by AUG on 2019/2/1.
//  Copyright © 2019年 JmoVxia. All rights reserved.
//

import UIKit

class CLPasswordInputViewConfig: NSObject {
    /// 密码的位数
    var passwordLength = 6
    /// 边框正方形的大小
    var squareSize = 45.0
    /// 边框圆角
    var cornerRadius = 5.0
    /// 边距相对中间间隙倍数
    var spaceRatio = 2.0
    /// 线条宽度
    var lineWidth = 5.0
    /// 圆点大小
    var dotRadius = 0.0
    /// 黑点颜色
    var dotColor = UIColor.black
    /// 边框颜色
    var borderColor = UIColor.lightGray
    /// 输入框背景颜色
    var inputBackgroundColor = UIColor.white
    /// 控件背景颜色
    var backgroundColor = UIColor.white
    /// 键盘类型
    var keyboardType = UIKeyboardType.asciiCapable
    /// 文字属性
    var attributes: [NSAttributedString.Key: Any]?
    /// 输入限制正则
    var regex = "^[a-zA-Z0-9]*$"
    /// 是否转化为全大写
    var shouldConvertToUpper = true
    /// 是否支持三方键盘
    var thirdPartyKeyboardEnabled = false

    fileprivate class func config() -> CLPasswordInputViewConfig {
        return CLPasswordInputViewConfig()
    }
}

protocol CLPasswordInputViewDelegate: AnyObject {
    /// 输入改变
    func passwordInputViewDidChange(passwordInputView: CLPasswordInputView) -> Void
    /// 点击删除
    func passwordInputViewDidDeleteBackward(passwordInputView: CLPasswordInputView) -> Void
    /// 输入完成
    func passwordInputViewCompleteInput(passwordInputView: CLPasswordInputView) -> Void
    /// 开始输入
    func passwordInputViewBeginInput(passwordInputView: CLPasswordInputView) -> Void
    /// 结束输入
    func passwordInputViewEndInput(passwordInputView: CLPasswordInputView) -> Void
}

extension CLPasswordInputViewDelegate {
    /// 输入改变
    func passwordInputViewDidChange(passwordInputView: CLPasswordInputView) {}

    /// 点击删除
    func passwordInputViewDidDeleteBackward(passwordInputView: CLPasswordInputView) {}

    /// 输入完成
    func passwordInputViewCompleteInput(passwordInputView: CLPasswordInputView) {}

    /// 开始输入
    func passwordInputViewBeginInput(passwordInputView: CLPasswordInputView) {}

    /// 结束输入
    func passwordInputViewEndInput(passwordInputView: CLPasswordInputView) {}
}

class CLPasswordInputView: UIView {
    override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = config.backgroundColor
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    weak var delegate: CLPasswordInputViewDelegate?

    private(set) var config = CLPasswordInputViewConfig.config()

    private(set) var text = String()

    private(set) var isShowKeyboard = false
}

extension CLPasswordInputView {
    override func becomeFirstResponder() -> Bool {
        if !isShowKeyboard {
            delegate?.passwordInputViewBeginInput(passwordInputView: self)
        }
        isShowKeyboard = true
        return super.becomeFirstResponder()
    }

    override func resignFirstResponder() -> Bool {
        if isShowKeyboard {
            delegate?.passwordInputViewEndInput(passwordInputView: self)
        }
        isShowKeyboard = false
        return super.resignFirstResponder()
    }

    override var canBecomeFirstResponder: Bool {
        true
    }

    override var canResignFirstResponder: Bool {
        true
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        guard !isFirstResponder else { return }
        _ = becomeFirstResponder()
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        setNeedsDisplay()
    }

    override func draw(_ rect: CGRect) {
        guard let context = UIGraphicsGetCurrentContext() else { return }
        let height = rect.size.height
        let width = rect.size.width
        let squareWidth = min(max(min(height, config.squareSize), config.dotRadius * 4), height)
        let middleSpace = CGFloat(width - CGFloat(config.passwordLength) * squareWidth) / CGFloat(CGFloat(config.passwordLength - 1) + config.spaceRatio * 2)
        let leftSpace = middleSpace * config.spaceRatio
        let y = (height - squareWidth) * 0.5

        let rectPaths = (0 ..< config.passwordLength).map { i in
            UIBezierPath(roundedRect: CGRect(x: leftSpace + CGFloat(i) * squareWidth + CGFloat(i) * middleSpace, y: y, width: squareWidth, height: squareWidth), cornerRadius: config.cornerRadius).cgPath
        }
        context.addPath(rectPaths.reduce(CGMutablePath()) { $0.addPath($1); return $0 })
        context.setLineWidth(config.lineWidth)
        context.setStrokeColor(config.borderColor.cgColor)
        context.setFillColor(config.inputBackgroundColor.cgColor)
        context.drawPath(using: .fillStroke)

        let dotPaths = text.enumerated().map { i, char -> CGPath in
            if config.dotRadius > .zero {
                return UIBezierPath(arcCenter: CGPoint(x: leftSpace + CGFloat(i + 1) * squareWidth + CGFloat(i) * middleSpace - squareWidth * 0.5, y: y + squareWidth * 0.5), radius: config.dotRadius, startAngle: 0, endAngle: .pi * 2, clockwise: true).cgPath
            } else {
                let string = String(char) as NSString
                let size = string.size(withAttributes: config.attributes)
                string.draw(at: CGPoint(x: leftSpace + CGFloat(i + 1) * squareWidth + CGFloat(i) * middleSpace - squareWidth * 0.5 - size.width * 0.5, y: y + squareWidth * 0.5 - size.height * 0.5), withAttributes: config.attributes)
                return CGMutablePath()
            }
        }
        context.addPath(dotPaths.reduce(CGMutablePath()) { $0.addPath($1); return $0 })
        context.setFillColor(config.dotColor.cgColor)
        context.drawPath(using: .fill)
    }
}

extension CLPasswordInputView {
    func updateConfig(_ configBlock: ((CLPasswordInputViewConfig) -> Void)?) {
        configBlock?(config)
        backgroundColor = config.backgroundColor
        setNeedsDisplay()
    }
}

extension CLPasswordInputView: UIKeyInput {
    var hasText: Bool {
        !text.isEmpty
    }

    func insertText(_ text: String) {
        guard text.range(of: config.regex, options: .regularExpression) != nil else { return }
        guard self.text.count < config.passwordLength else { return }

        let text = config.shouldConvertToUpper ? text.uppercased() : text
        self.text.append(text)
        setNeedsDisplay()

        delegate?.passwordInputViewDidChange(passwordInputView: self)
        guard self.text.count == config.passwordLength else { return }
        delegate?.passwordInputViewCompleteInput(passwordInputView: self)
    }

    func deleteBackward() {
        guard !text.isEmpty else { return }
        text.removeLast()
        delegate?.passwordInputViewDidChange(passwordInputView: self)
        delegate?.passwordInputViewDidDeleteBackward(passwordInputView: self)
        setNeedsDisplay()
    }

    var keyboardType: UIKeyboardType {
        get {
            return config.keyboardType
        }
        set {}
    }

    var isSecureTextEntry: Bool {
        get {
            return !config.thirdPartyKeyboardEnabled
        }
        set {}
    }
}
